pmd-java/src/main/resources/category/java/bestpractices.xml

```diff
@@ -802,6 +802,76 @@ public class MyTest {
         </example>
     </rule>
 
+    <rule name="JUnit5TestShouldBePackagePrivate"
+          language="java"
+          since="6.35"
+          message="JUnit 5 tests that use @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest should be package private."
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          typeResolution="true"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_bestpractices.html#junit5testshouldbepackageprivate">
+        <description>
+            JUnit 5 tests should be package private at the class level and for each method that uses @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="version" value="2.0"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//ClassOrInterfaceDeclaration[
+       matches(@SimpleName, $testClassPattern) and
+       @Abstract=false() and
+       @PackagePrivate=false()]
+       [ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration
+        [MethodDeclaration]
+        [Annotation//Name[
+            pmd-java:typeIs('org.junit.jupiter.api.Test') or pmd-java:typeIs('org.junit.jupiter.api.RepeatedTest')
+            or pmd-java:typeIs('org.junit.jupiter.api.TestFactory') or pmd-java:typeIs('org.junit.jupiter.api.TestTemplate')
+            or pmd-java:typeIs('org.junit.jupiter.params.ParameterizedTest')
+        ]]]
+|
+//ClassOrInterfaceDeclaration[
+       matches(@SimpleName, $testClassPattern)]
+    /ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration
+        [MethodDeclaration[@PackagePrivate=false()]]
+        [Annotation//Name[
+            pmd-java:typeIs('org.junit.jupiter.api.Test') or pmd-java:typeIs('org.junit.jupiter.api.RepeatedTest')
+            or pmd-java:typeIs('org.junit.jupiter.api.TestFactory') or pmd-java:typeIs('org.junit.jupiter.api.TestTemplate')
+            or pmd-java:typeIs('org.junit.jupiter.params.ParameterizedTest')
+        ]]
+]]>
+                </value>
+            </property>
+            <property name="testClassPattern" type="Regex" description="The regex pattern used to identify test classes" value="Test" />
```

Comment:
We're looking for a junit 5 test class, ie a class in which some method has one of the test annotations mentioned above. I don't think there's a need to check for the name of the class?

Reply:
Well I followed what was in `JUnit4TestShouldUseTestAnnotation` but I agree I don't see the rationale for it.